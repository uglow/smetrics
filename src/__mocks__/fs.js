'use strict';

const path = require('path');

const fs = jest.createMockFromModule('fs');

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = Object.create(null);

function __setMockFiles(newMockFiles) {
  mockFiles = Object.create(null);
  for (const file in newMockFiles) {
    const dir = path.dirname(file);

    if (!mockFiles[dir]) {
      mockFiles[dir] = [];
    }
    mockFiles[dir].push(path.basename(file));
  }
}

function resetMock() {
  mockFiles = {};
}

// Write custom mock functions

function existsSync(filePath) {
  return !!mockFiles[filePath];
}

function readFileSync(filePath) {
  return mockFiles[filePath];
}

function writeFileSync(filePath, data) {
  mockFiles[filePath] = data;
}

function unlinkSync(filePath) {
  delete mockFiles[filePath];
}

fs.__setMockFiles = __setMockFiles;
fs.existsSync = existsSync;
fs.readFileSync = readFileSync;
fs.writeFileSync = writeFileSync;
fs.unlinkSync = unlinkSync;

fs.__resetMock = resetMock;

module.exports = fs;
