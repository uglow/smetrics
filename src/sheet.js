const { JWT } = require('google-auth-library');

const GOOGLE_AUTH_SCOPES = ['https://www.googleapis.com/auth/spreadsheets'];
const SHEETS_API = 'https://sheets.googleapis.com/v4/spreadsheets/';

const MONTH = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

const DATE_FORMATS = {
  milliseconds: (dateMillis) => dateMillis,
  googleDate: (dateMillis) => {
    const d = new Date(dateMillis);
    return (
      [d.getUTCDate(), MONTH[d.getUTCMonth()], d.getUTCFullYear()].join('-') +
      ' ' +
      [d.getUTCHours(), String(d.getUTCMinutes()).padStart(2, '0'), String(d.getSeconds()).padStart(2, '0')].join(':')
    );
  },
};

class SheetService {
  constructor(spreadsheetId, { dateFormat = 'milliseconds' } = {}) {
    if (!spreadsheetId) throw new Error('Spreadsheet id must be supplied');
    this.spreadsheetId = spreadsheetId;

    const dateFormatType = typeof dateFormat;

    if (!['string', 'function'].includes(dateFormatType)) {
      throw new Error(`Invalid dateFormat. dateFormat must be a string or a function`);
    } else if (dateFormatType === 'string') {
      if (!DATE_FORMATS[dateFormat]) {
        throw new Error(`Invalid dateFormat. dateFormat must be one of: ${Object.keys(DATE_FORMATS).join(', ')}`);
      }
      this.dateFormat = DATE_FORMATS[dateFormat];
    } else {
      this.dateFormat = dateFormat;
    }
  }

  authorize(creds, authService = JWT) {
    this.client = new authService({ email: creds.clientEmail, key: creds.privateKey, scopes: GOOGLE_AUTH_SCOPES });
    return this.client.authorize(); // Returns the access token
  }

  async addData(records) {
    // First step. Assume we have multi-dimensional array.
    records.forEach((recordArr) => {
      // Start by grouping the elements by category
      const groupedItems = recordArr.reduce(groupBy('sheet'), {});

      // Now for each sheet...
      Object.keys(groupedItems).forEach(async (sheet) => {
        // Create the sheet (in case it does not exist)
        try {
          await this.createSheet(sheet);
        } catch (err) {
          if (err && err.errors && err.errors.length) {
            if (!err.errors[0].message.includes('already exists. Please enter another name')) {
              err.errors.forEach((item) => console.log(item));
            }
          } else {
            console.error(err);
          }
        }

        // Update the header then add a new row, appending the dateTime as the first row
        await this.updateHeader(sheet, ['DateTime'].concat(groupedItems[sheet].map((item) => item.metric))).then(() =>
          this.appendData(
            sheet,
            [this.dateFormat(groupedItems[sheet][0].t)].concat(groupedItems[sheet].map((item) => item.value)),
          ),
        );
      });
    });
  }

  /**
   * Update the header, with the first column being the timestamp
   * @param sheetName
   * @param values
   * @returns {Promise<*>}
   */
  async updateHeader(sheetName, values) {
    const range = `'${sheetName}'!A1:${columnToLetter(values.length)}1`; // +1 is the DateTime generated-column
    const url = `${SHEETS_API}${this.spreadsheetId}/values/${range}?access_token=${this.client.credentials.access_token}&valueInputOption=RAW&includeValuesInResponse=true`;
    const data = {
      range,
      majorDimension: 'ROWS',
      values: [values],
    };
    const res = await this.client.request({ url, method: 'put', data });
    return res.data;
  }

  async appendData(sheetName, values) {
    const range = `'${sheetName}'!A1:${columnToLetter(values.length)}1`; // +1 is the DateTime generated-column
    const url = `${SHEETS_API}${this.spreadsheetId}/values/${range}:append?access_token=${this.client.credentials.access_token}&valueInputOption=USER_ENTERED&insertDataOption=INSERT_ROWS`;
    const data = {
      range,
      majorDimension: 'ROWS',
      values: [values],
    };

    const res = await this.client.request({ url, method: 'post', data });
    return res.data;
  }

  async createSheet(sheetName) {
    const url = `${SHEETS_API}${this.spreadsheetId}:batchUpdate?access_token=${this.client.credentials.access_token}`;
    const data = {
      requests: [
        {
          addSheet: {
            properties: {
              title: sheetName,
            },
          },
        },
      ],
    };
    return this.client.request({ url, method: 'post', data });
  }
}

/**
 * Higher-order function which can be used by array.reduce(groupBy('key'), {}) to convert
 * an array into an object grouped-by the specified key
 * @param {string} keyProperty          Property to group the elements on
 * @returns {function(acc, curr): {}}   Accumulator function for use within array.reduce()
 */
function groupBy(keyProperty) {
  return (acc, curr) =>
    Object.assign(
      acc,
      { [curr[keyProperty]]: [...(acc[curr[keyProperty]] || []), curr] }, // Uses array-destructuring to concat to existing or create new arr
    );
}

function columnToLetter(column) {
  let temp,
    letter = '';
  while (column > 0) {
    temp = (column - 1) % 26;
    letter = String.fromCharCode(temp + 65) + letter;
    column = (column - temp - 1) / 26;
  }
  return letter;
}

module.exports = {
  SheetService,
};
