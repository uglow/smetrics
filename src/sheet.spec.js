import { SheetService } from './sheet';
import fs from 'fs';
import { fileURLToPath } from 'url';
import { jest } from '@jest/globals';

describe('SheetService', () => {
  function getMockClient(mockResponse = '') {
    return {
      credentials: {
        access_token: 'token',
      },
      request: jest.fn().mockResolvedValue({ data: mockResponse }),
    };
  }

  it('should be createable using a non-blank spreadsheet id', () => {
    expect(() => new SheetService()).toThrow('Spreadsheet id must be supplied');

    const service = new SheetService('foo');
    expect(service).toBeDefined();
  });

  describe('dateFormat', () => {
    it('should be possible to specify the dateFormat as "googleDate" or "milliseconds"', () => {
      let service = new SheetService('foo', { dateFormat: 'googleDate' });
      expect(service.dateFormat(123)).toEqual('1-Jan-1970 0:00:00');

      service = new SheetService('foo', { dateFormat: 'milliseconds' });
      expect(service.dateFormat(123)).toEqual(123);

      expect(() => new SheetService('foo', { dateFormat: 'unknown' })).toThrow(
        'Invalid dateFormat. dateFormat must be one of: milliseconds, googleDate',
      );
    });

    it('should be possible to specify the dateFormat as a function', () => {
      const myDateFunction = () => 'not a date!';

      let service = new SheetService('foo', { dateFormat: myDateFunction });
      expect(service.dateFormat(123)).toEqual('not a date!');

      expect(() => new SheetService('foo', { dateFormat: 123 })).toThrow(
        'Invalid dateFormat. dateFormat must be a string or a function',
      );
    });
  });

  describe('.authorize()', () => {
    it('should be call the clientService.authorize() method, passing the credentials and scopes', () => {
      const service = new SheetService('foo');
      const mockAuthService = function (...params) {
        return {
          authorize() {},
          params: params,
        };
      };

      service.authorize(
        {
          clientEmail: 'a@b.com',
          privateKey: 'really long key',
        },
        mockAuthService,
      );

      // The client was called with the following params
      expect(service.client.params).toEqual([
        {
          email: 'a@b.com',
          key: 'really long key',
          scopes: ['https://www.googleapis.com/auth/spreadsheets'],
        },
      ]);
    });
  });

  describe('.updateHeader()', () => {
    it('should be call the clientService.request() method, passing the correct data', async () => {
      const service = new SheetService('foo');
      service.client = getMockClient();

      await service.updateHeader('Sheet 2', ['DateTime', 'Metric 1', 'Metric 2']);

      expect(service.client.request.mock.calls).toMatchSnapshot();
    });
  });

  describe('.appendData()', () => {
    it('should be call the clientService.request() method, passing the correct data', async () => {
      const service = new SheetService('foo');
      service.client = getMockClient();

      await service.appendData('Sheet 2', ['12322311277', '78', 12]);

      expect(service.client.request.mock.calls).toMatchSnapshot();
    });
  });

  describe('.createSheet()', () => {
    it('should be call the clientService.request() method, passing the correct data', async () => {
      const service = new SheetService('foo');
      service.client = getMockClient();

      await service.createSheet('Sp a ces');

      expect(service.client.request.mock.calls).toMatchSnapshot();
    });
  });

  describe('.addData()', () => {
    it('should call createSheet(), updateHeader() and appendData() for each record it is passed', async () => {
      const service = new SheetService('foo');
      service.createSheet = jest.fn().mockResolvedValue();
      service.updateHeader = jest.fn().mockResolvedValue();
      service.appendData = jest.fn().mockResolvedValue();
      const fixture = fileURLToPath(new URL('../fixtures/smetrics.json', import.meta.url));
      const data = JSON.parse(fs.readFileSync(fixture, 'utf8'));

      await service.addData(data);

      expect(service.createSheet.mock.calls[0]).toEqual(['sheet a']);
      expect(service.createSheet.mock.calls[1]).toEqual(['sheet b']);

      expect(service.updateHeader.mock.calls[0]).toEqual(['sheet a', ['DateTime', 'Metric 1', 'Metric 2']]);
      expect(service.updateHeader.mock.calls[1]).toEqual(['sheet b', ['DateTime', 'Objects of the future']]);

      expect(service.appendData.mock.calls[0]).toMatchSnapshot();
      expect(service.appendData.mock.calls[1]).toMatchSnapshot();
    });

    it('should apply the dateFormat correctly when calling appendData()', async () => {
      const service = new SheetService('foo', { dateFormat: 'googleDate' });
      service.createSheet = jest.fn().mockResolvedValue();
      service.updateHeader = jest.fn().mockResolvedValue();
      service.appendData = jest.fn().mockResolvedValue();
      const fixture = fileURLToPath(new URL('../fixtures/smetrics.json', import.meta.url));
      const data = JSON.parse(fs.readFileSync(fixture, 'utf8'));

      await service.addData(data);

      expect(service.createSheet.mock.calls[0]).toEqual(['sheet a']);
      expect(service.createSheet.mock.calls[1]).toEqual(['sheet b']);

      expect(service.updateHeader.mock.calls[0]).toEqual(['sheet a', ['DateTime', 'Metric 1', 'Metric 2']]);
      expect(service.updateHeader.mock.calls[1]).toEqual(['sheet b', ['DateTime', 'Objects of the future']]);

      expect(service.appendData.mock.calls[0]).toMatchSnapshot();
      expect(service.appendData.mock.calls[1]).toMatchSnapshot();
    });

    it('should still add data even if createSheet() rejects (as happens when the sheet exists already)', async () => {
      const service = new SheetService('foo');
      service.createSheet = jest.fn().mockRejectedValue({
        errors: [
          {
            message:
              'Invalid requests[0].addSheet: A sheet with the name "sheet a" already exists. Please enter another name.',
          },
        ],
      });
      service.updateHeader = jest.fn().mockResolvedValue();
      service.appendData = jest.fn().mockResolvedValue();
      const fixture = fileURLToPath(new URL('../fixtures/smetrics.json', import.meta.url));
      const data = JSON.parse(fs.readFileSync(fixture, 'utf8'));

      await service.addData(data);

      expect(service.createSheet.mock.calls.length).toEqual(2);
      expect(service.updateHeader.mock.calls.length).toEqual(2);
      expect(service.appendData.mock.calls.length).toEqual(2);
    });

    it('should be able to process two-or-more records at a time', async () => {
      const service = new SheetService('foo');
      service.createSheet = jest.fn().mockResolvedValue();
      service.updateHeader = jest.fn().mockResolvedValue();
      service.appendData = jest.fn().mockResolvedValue();
      const fixture = fileURLToPath(new URL('../fixtures/smetrics-multiple.json', import.meta.url));
      const data = JSON.parse(fs.readFileSync(fixture, 'utf8'));

      await service.addData(data);

      expect(service.createSheet.mock.calls[0]).toEqual(['sheet a']);
      expect(service.createSheet.mock.calls[1]).toEqual(['sheet b']);

      expect(service.updateHeader.mock.calls[0]).toEqual(['sheet a', ['DateTime', 'Metric 1']]);
      expect(service.updateHeader.mock.calls[1]).toEqual(['sheet b', ['DateTime', 'Objects of the future']]);

      expect(service.appendData.mock.calls[0]).toMatchSnapshot();
      expect(service.appendData.mock.calls[1]).toMatchSnapshot();
    });
  });
});
