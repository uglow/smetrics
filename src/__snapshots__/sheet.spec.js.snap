// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SheetService .addData() should be able to process two-or-more records at a time 1`] = `
Array [
  "sheet a",
  Array [
    1536627335859,
    11,
  ],
]
`;

exports[`SheetService .addData() should be able to process two-or-more records at a time 2`] = `
Array [
  "sheet b",
  Array [
    1536627335862,
    "car",
  ],
]
`;

exports[`SheetService .addData() should call createSheet(), updateHeader() and appendData() for each record it is passed 1`] = `
Array [
  "sheet a",
  Array [
    1536627335859,
    11,
    "503",
  ],
]
`;

exports[`SheetService .addData() should call createSheet(), updateHeader() and appendData() for each record it is passed 2`] = `
Array [
  "sheet b",
  Array [
    1536627335862,
    "car",
  ],
]
`;

exports[`SheetService .appendData() should be call the clientService.request() method, passing the correct data 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "majorDimension": "ROWS",
        "range": "'Sheet 2'!A1:C1",
        "values": Array [
          Array [
            "12322311277",
            "78",
            12,
          ],
        ],
      },
      "method": "post",
      "url": "https://sheets.googleapis.com/v4/spreadsheets/foo/values/'Sheet 2'!A1:C1:append?access_token=token&valueInputOption=RAW&insertDataOption=INSERT_ROWS",
    },
  ],
]
`;

exports[`SheetService .createSheet() should be call the clientService.request() method, passing the correct data 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "requests": Array [
          Object {
            "addSheet": Object {
              "properties": Object {
                "title": "Sp a ces",
              },
            },
          },
        ],
      },
      "method": "post",
      "url": "https://sheets.googleapis.com/v4/spreadsheets/foo:batchUpdate?access_token=token",
    },
  ],
]
`;

exports[`SheetService .updateHeader() should be call the clientService.request() method, passing the correct data 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "majorDimension": "ROWS",
        "range": "'Sheet 2'!A1:C1",
        "values": Array [
          Array [
            "DateTime",
            "Metric 1",
            "Metric 2",
          ],
        ],
      },
      "method": "put",
      "url": "https://sheets.googleapis.com/v4/spreadsheets/foo/values/'Sheet 2'!A1:C1?access_token=token&valueInputOption=RAW&includeValuesInResponse=true",
    },
  ],
]
`;
